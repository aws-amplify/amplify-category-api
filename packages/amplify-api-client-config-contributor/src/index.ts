export { ApiClientConfigContributor } from './api-client-config-contributor';
export { ApiClientConfig, ApiClientConfigMapping } from './api-client-config';
