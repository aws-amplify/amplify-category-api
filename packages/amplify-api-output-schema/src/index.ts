import { z } from 'zod';
import { apiOutputSchema as apiOutputSchemaV1 } from './v1';

export const versionedApiOutputSchema = z.discriminatedUnion('version', [
  apiOutputSchemaV1,
  // this is where additional api major version schemas would go
]);

/**
 *
 */
export type ApiOutput = z.infer<typeof versionedApiOutputSchema>;

export const ApiOutputKey = 'apiOutput';
