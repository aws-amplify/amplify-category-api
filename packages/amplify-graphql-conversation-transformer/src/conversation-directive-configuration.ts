import { DataSourceProvider } from '@aws-amplify/graphql-transformer-interfaces';
import { DirectiveNode, FieldDefinitionNode, InputObjectTypeDefinitionNode, ObjectTypeDefinitionNode } from 'graphql';
import { ConversationModel } from './graphql-types/conversation-model';
import { MessageModel } from './graphql-types/message-model';
import { Tool } from './tools/process-tools';

/**
 * Configuration for the Conversation Directive
 */
export type ConversationDirectiveConfiguration = {
  // From the GraphQL Schema
  parent: ObjectTypeDefinitionNode;
  field: FieldDefinitionNode;
  directive: DirectiveNode;

  // Raw Directive Input
  aiModel: string;
  systemPrompt: string;
  inferenceConfiguration: ConversationInferenceConfiguration | undefined;
  tools: ToolDefinition[];
  /**
   * Custom handler function name.
   *
   * @deprecated Replaced by 'handler'
   */
  functionName: string | undefined;
  handler: ConversationHandlerFunctionConfiguration | undefined;

  // Generated within the Conversation Transformer
  toolSpec: Tool[] | undefined;
  conversation: ConversationModel;
  message: MessageModel;
  assistantResponseMutation: { field: FieldDefinitionNode; input: InputObjectTypeDefinitionNode };
  assistantResponseStreamingMutation: { field: FieldDefinitionNode; input: InputObjectTypeDefinitionNode };
  assistantResponseSubscriptionField: FieldDefinitionNode;
  dataSources: ConversationDirectiveDataSources;
};

/**
 * Conversation Directive Handler Function Configuration
 */
export type ConversationHandlerFunctionConfiguration = {
  functionName: string;
  eventVersion: string;
};

/**
 * Conversation Directive Inference Configuration
 */
export type ConversationInferenceConfiguration = {
  maxTokens?: number;
  temperature?: number;
  topP?: number;
};

/**
 * Conversation Directive Tool Definition
 */
export type ToolDefinition = {
  name: string;
  description: string;
};

/**
 * Conversation Directive Data Sources
 */
export type ConversationDirectiveDataSources = {
  lambdaFunctionDataSource: DataSourceProvider;
  messageTableDataSource: DataSourceProvider;
  conversationTableDataSource: DataSourceProvider;
};
