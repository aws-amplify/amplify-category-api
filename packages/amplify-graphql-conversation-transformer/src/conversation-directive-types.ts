import { DataSourceProvider } from '@aws-amplify/graphql-transformer-interfaces';
import { DirectiveNode, FieldDefinitionNode, InputObjectTypeDefinitionNode, ObjectTypeDefinitionNode } from 'graphql';
import { MessageModel } from './graphql-types/message-model';
import { ConversationModel } from './graphql-types/conversation-model';
import { Tool } from './utils/tools';

export const LIST_MESSAGES_INDEX_NAME = 'gsi-ConversationMessage.conversationId.createdAt';

/**
 * Configuration for the Conversation Directive
 */
export type ConversationDirectiveConfiguration = {
  // From the GraphQL Schema
  parent: ObjectTypeDefinitionNode;
  field: FieldDefinitionNode;
  directive: DirectiveNode;

  // Raw Directive Input
  aiModel: string;
  systemPrompt: string;
  inferenceConfiguration: ConversationInferenceConfiguration | undefined;
  tools: ToolDefinition[];
  functionName: string | undefined;
  handler: ConversationHandlerFunctionConfiguration | undefined;

  // Generated within the Conversation Transformer
  toolSpec: Tool[] | undefined;
  conversation: ConversationModel;
  message: MessageModel;
  assistantResponseMutation: { field: FieldDefinitionNode; input: InputObjectTypeDefinitionNode };
  assistantResponseSubscriptionField: FieldDefinitionNode;
  dataSources: {
    conversationTable: DataSourceProvider;
    messageTable: DataSourceProvider;
    lambdaFunction: DataSourceProvider;
  };
};

/**
 * Conversation Handler Function Configuration
 */
export type ConversationHandlerFunctionConfiguration = {
  functionName: string;
  eventVersion: string;
};

/**
 * Conversation Inference Configuration
 */
export type ConversationInferenceConfiguration = {
  maxTokens?: number;
  temperature?: number;
  topP?: number;
};

export type ToolDefinition = {
  name: string;
  description: string;
};
