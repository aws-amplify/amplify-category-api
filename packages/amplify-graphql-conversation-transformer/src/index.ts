export { ConversationTransformer } from './grapqhl-conversation-transformer';