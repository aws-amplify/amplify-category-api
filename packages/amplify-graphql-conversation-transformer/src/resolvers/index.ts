import { assistantResponsePipelineDefinition } from './assistant-response-pipeline-definition';
import { assistantResponseSubscriptionPipelineDefinition } from './assistant-response-subscription-pipeline-definition';
import { generateResolverFunction, generateResolverPipeline } from './generate-resolver';
import { listMessagesInitFunctionDefinition } from './list-messages-init-resolver';
import { sendMessagePipelineDefinition } from './send-message-pipeline-definition';

export {
  assistantResponsePipelineDefinition,
  assistantResponseSubscriptionPipelineDefinition,
  generateResolverFunction,
  generateResolverPipeline,
  listMessagesInitFunctionDefinition,
  sendMessagePipelineDefinition,
};
