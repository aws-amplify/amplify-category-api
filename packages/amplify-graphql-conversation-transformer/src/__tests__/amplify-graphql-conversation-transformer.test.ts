import { AuthTransformer } from '@aws-amplify/graphql-auth-transformer';
import { IndexTransformer, PrimaryKeyTransformer } from '@aws-amplify/graphql-index-transformer';
import { ModelTransformer } from '@aws-amplify/graphql-model-transformer';
import { DDB_AMPLIFY_MANAGED_DATASOURCE_STRATEGY, GraphQLTransform, validateModelSchema } from '@aws-amplify/graphql-transformer-core';
import { AppSyncAuthConfiguration } from '@aws-amplify/graphql-transformer-interfaces';
import { testTransform } from '@aws-amplify/graphql-transformer-test-utils';
import { parse, print } from 'graphql';
import { ConversationTransformer } from '..';
import { BelongsToTransformer, HasManyTransformer, HasOneTransformer } from '@aws-amplify/graphql-relational-transformer';
import { FunctionTransformer } from '../../../amplify-graphql-function-transformer/src';

test('conversation route valid schema', () => {
  const authConfig: AppSyncAuthConfiguration = {
    defaultAuthentication: {
      authenticationType: 'AMAZON_COGNITO_USER_POOLS',
    },
    additionalAuthenticationProviders: [],
  };

  const inputSchema = `
    type Foo @model {
      bar: Int
    }

    type Mutation {
        pirateChat(sessionId: ID, message: String): String
        @function(name: "abc")
        @conversation(aiModel: "Claude3Haiku")
    }
  `;

  // angryChat(sessionId: ID, message: String): String
        // @conversation(aiModel: "Claude3Haiku")

  // const modelTransformer = new ModelTransformer();
  // const authTransformer = new AuthTransformer();
  // const hasManyTransformer = new HasManyTransformer();
  // const belongsToTransformer = new BelongsToTransformer();

  // const transformers = [
  //   modelTransformer,
  //   new PrimaryKeyTransformer(),
  //   new IndexTransformer(),
  //   hasManyTransformer,
  //   belongsToTransformer,
  //   new ConversationTransformer(modelTransformer, authTransformer),
  //   authTransformer,
  // ];

  const modelTransformer = new ModelTransformer();
  const authTransformer = new AuthTransformer();
  const indexTransformer = new IndexTransformer();
  const hasOneTransformer = new HasOneTransformer();
  const belongsToTransformer = new BelongsToTransformer();
  const hasManyTransformer = new HasManyTransformer();
  // The default list of transformers should match DefaultDirectives in packages/amplify-graphql-directives/src/index.ts
  const transformers = [
    modelTransformer,
    new FunctionTransformer(),
    new PrimaryKeyTransformer(),
    indexTransformer,
    hasManyTransformer,
    hasOneTransformer,
    belongsToTransformer,
    new ConversationTransformer(modelTransformer, hasManyTransformer, belongsToTransformer, authTransformer),
    authTransformer,
  ]

  // const processedDocumentNode  = new GraphQLTransform({ authConfig, transformers }).preProcessSchema(parse(inputSchema));
  // const processedSchema = print(processedDocumentNode);
  // expect(processedSchema).toMatchSnapshot();

  const out = testTransform({
    schema: inputSchema,
    authConfig,
    transformers,
    dataSourceStrategies: {
      'Foo': DDB_AMPLIFY_MANAGED_DATASOURCE_STRATEGY
    }
  });

  expect(out).toBeDefined();
  const schema = parse(out.schema);
  validateModelSchema(schema);
  expect(out.schema).toMatchSnapshot();
});


// test('conversation route valid schema', () => {
//   const authConfig: AppSyncAuthConfiguration = {
//     defaultAuthentication: {
//       authenticationType: 'AMAZON_COGNITO_USER_POOLS',
//     },
//     additionalAuthenticationProviders: [],
//   };

//   const inputSchema = `
//     type ConversationMessage_pirateChat
//     @model
//     @auth(rules: [{allow: owner, ownerField: "owner"}])
//     {
//         conversationSessionId: ID!
//         session: ConversationSession_pirateChat @belongsTo(references: "conversationSessionId")
//         sender: ConversationEventSender
//         text: String
//     }

//     type ConversationSession_pirateChat
//     @model
//     @auth(rules: [{allow: owner, ownerField: "owner"}])
//     {
//         events: [ConversationMessage_pirateChat] @hasMany(references: "conversationSessionId")
//     }

//     enum ConversationEventSender {
//         user
//         assistant
//     }

//     type Mutation {
//         pirateChat(sessionId: ID, message: String): String
//         @conversation(
//             aiModel: "Claude3Haiku",
//             sessionModel: { name: "ConversationSession_pirateChat" },
//             eventModel: { name: "ConversationMessage_pirateChat" }
//         )
//     }
//   `;

//   const out = testTransform({
//     schema: inputSchema,
//     authConfig,
//     transformers: [
//       new ModelTransformer(),
//       new AuthTransformer(),
//       new PrimaryKeyTransformer(),
//       new HasManyTransformer(),
//       new BelongsToTransformer(),
//       new ConversationTransformer(),
//     ],
//   });

//   expect(out).toBeDefined();
//   const schema = parse(out.schema);
//   validateModelSchema(schema);
//   expect(out.schema).toMatchSnapshot();
// });
