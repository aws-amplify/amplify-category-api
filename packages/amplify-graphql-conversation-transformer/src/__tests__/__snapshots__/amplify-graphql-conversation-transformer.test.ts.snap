// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`conversation route valid schema 1`] = `
"type Foo {
  bar: Int
  id: ID!
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
}

type Mutation {
  pirateChat(sessionId: ID, message: String): String
  createFoo(input: CreateFooInput!, condition: ModelFooConditionInput): Foo
  updateFoo(input: UpdateFooInput!, condition: ModelFooConditionInput): Foo
  deleteFoo(input: DeleteFooInput!, condition: ModelFooConditionInput): Foo
  createConversationSessionpirateChat(input: CreateConversationSessionpirateChatInput!, condition: ModelConversationSessionpirateChatConditionInput): ConversationSessionpirateChat
  deleteConversationSessionpirateChat(input: DeleteConversationSessionpirateChatInput!, condition: ModelConversationSessionpirateChatConditionInput): ConversationSessionpirateChat
  createConversationMessagepirateChat(input: CreateConversationMessagepirateChatInput!, condition: ModelConversationMessagepirateChatConditionInput): ConversationMessagepirateChat
  deleteConversationMessagepirateChat(input: DeleteConversationMessagepirateChatInput!, condition: ModelConversationMessagepirateChatConditionInput): ConversationMessagepirateChat
}

enum ConversationMessageSender {
  user
  assistant
}

type ConversationSessionpirateChat {
  id: ID!
  name: String
  metadata: AWSJSON
  messages(filter: ModelConversationMessagepirateChatFilterInput, sortDirection: ModelSortDirection, limit: Int, nextToken: String): ModelConversationMessagepirateChatConnection
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
  owner: String
}

type ConversationMessagepirateChat {
  id: ID!
  conversationSessionId: ID!
  session: ConversationSessionpirateChat
  sender: ConversationMessageSender
  content: String
  context: AWSJSON
  uiComponents: [AWSJSON]
  createdAt: AWSDateTime!
  updatedAt: AWSDateTime!
  owner: String
}

input ModelStringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  contains: String
  notContains: String
  between: [String]
  beginsWith: String
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
  size: ModelSizeInput
}

input ModelIntInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelFloatInput {
  ne: Float
  eq: Float
  le: Float
  lt: Float
  ge: Float
  gt: Float
  between: [Float]
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelBooleanInput {
  ne: Boolean
  eq: Boolean
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
}

input ModelIDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  contains: ID
  notContains: ID
  between: [ID]
  beginsWith: ID
  attributeExists: Boolean
  attributeType: ModelAttributeTypes
  size: ModelSizeInput
}

input ModelSubscriptionStringInput {
  ne: String
  eq: String
  le: String
  lt: String
  ge: String
  gt: String
  contains: String
  notContains: String
  between: [String]
  beginsWith: String
  in: [String]
  notIn: [String]
}

input ModelSubscriptionIntInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
  in: [Int]
  notIn: [Int]
}

input ModelSubscriptionFloatInput {
  ne: Float
  eq: Float
  le: Float
  lt: Float
  ge: Float
  gt: Float
  between: [Float]
  in: [Float]
  notIn: [Float]
}

input ModelSubscriptionBooleanInput {
  ne: Boolean
  eq: Boolean
}

input ModelSubscriptionIDInput {
  ne: ID
  eq: ID
  le: ID
  lt: ID
  ge: ID
  gt: ID
  contains: ID
  notContains: ID
  between: [ID]
  beginsWith: ID
  in: [ID]
  notIn: [ID]
}

enum ModelAttributeTypes {
  binary
  binarySet
  bool
  list
  map
  number
  numberSet
  string
  stringSet
  _null
}

input ModelSizeInput {
  ne: Int
  eq: Int
  le: Int
  lt: Int
  ge: Int
  gt: Int
  between: [Int]
}

enum ModelSortDirection {
  ASC
  DESC
}

type ModelFooConnection {
  items: [Foo]!
  nextToken: String
}

input ModelFooFilterInput {
  bar: ModelIntInput
  id: ModelIDInput
  createdAt: ModelStringInput
  updatedAt: ModelStringInput
  and: [ModelFooFilterInput]
  or: [ModelFooFilterInput]
  not: ModelFooFilterInput
}

type Query {
  getFoo(id: ID!): Foo
  listFoos(filter: ModelFooFilterInput, limit: Int, nextToken: String): ModelFooConnection
  getConversationSessionpirateChat(id: ID!): ConversationSessionpirateChat
  listConversationSessionpirateChats(filter: ModelConversationSessionpirateChatFilterInput, limit: Int, nextToken: String): ModelConversationSessionpirateChatConnection
  getConversationMessagepirateChat(id: ID!): ConversationMessagepirateChat
  listConversationMessagepirateChats(filter: ModelConversationMessagepirateChatFilterInput, limit: Int, nextToken: String): ModelConversationMessagepirateChatConnection
}

input ModelFooConditionInput {
  bar: ModelIntInput
  and: [ModelFooConditionInput]
  or: [ModelFooConditionInput]
  not: ModelFooConditionInput
  createdAt: ModelStringInput
  updatedAt: ModelStringInput
}

input CreateFooInput {
  bar: Int
  id: ID
}

input UpdateFooInput {
  bar: Int
  id: ID!
}

input DeleteFooInput {
  id: ID!
}

input ModelSubscriptionFooFilterInput {
  bar: ModelSubscriptionIntInput
  id: ModelSubscriptionIDInput
  createdAt: ModelSubscriptionStringInput
  updatedAt: ModelSubscriptionStringInput
  and: [ModelSubscriptionFooFilterInput]
  or: [ModelSubscriptionFooFilterInput]
}

type Subscription {
  onCreateFoo(filter: ModelSubscriptionFooFilterInput): Foo @aws_subscribe(mutations: [\\"createFoo\\"])
  onUpdateFoo(filter: ModelSubscriptionFooFilterInput): Foo @aws_subscribe(mutations: [\\"updateFoo\\"])
  onDeleteFoo(filter: ModelSubscriptionFooFilterInput): Foo @aws_subscribe(mutations: [\\"deleteFoo\\"])
  onCreateConversationMessagepirateChat(filter: ModelSubscriptionConversationMessagepirateChatFilterInput, owner: String): ConversationMessagepirateChat @aws_subscribe(mutations: [\\"createConversationMessagepirateChat\\"])
}

type ModelConversationSessionpirateChatConnection {
  items: [ConversationSessionpirateChat]!
  nextToken: String
}

input ModelConversationSessionpirateChatFilterInput {
  id: ModelIDInput
  name: ModelStringInput
  metadata: ModelStringInput
  createdAt: ModelStringInput
  updatedAt: ModelStringInput
  and: [ModelConversationSessionpirateChatFilterInput]
  or: [ModelConversationSessionpirateChatFilterInput]
  not: ModelConversationSessionpirateChatFilterInput
  owner: ModelStringInput
}

input ModelConversationSessionpirateChatConditionInput {
  name: ModelStringInput
  metadata: ModelStringInput
  and: [ModelConversationSessionpirateChatConditionInput]
  or: [ModelConversationSessionpirateChatConditionInput]
  not: ModelConversationSessionpirateChatConditionInput
  createdAt: ModelStringInput
  updatedAt: ModelStringInput
  owner: ModelStringInput
}

input CreateConversationSessionpirateChatInput {
  id: ID
  name: String
  metadata: AWSJSON
}

input DeleteConversationSessionpirateChatInput {
  id: ID!
}

type ModelConversationMessagepirateChatConnection {
  items: [ConversationMessagepirateChat]!
  nextToken: String
}

input ModelConversationMessageSenderInput {
  eq: ConversationMessageSender
  ne: ConversationMessageSender
}

input ModelConversationMessagepirateChatFilterInput {
  id: ModelIDInput
  conversationSessionId: ModelIDInput
  sender: ModelConversationMessageSenderInput
  content: ModelStringInput
  context: ModelStringInput
  uiComponents: ModelStringInput
  createdAt: ModelStringInput
  updatedAt: ModelStringInput
  and: [ModelConversationMessagepirateChatFilterInput]
  or: [ModelConversationMessagepirateChatFilterInput]
  not: ModelConversationMessagepirateChatFilterInput
  owner: ModelStringInput
}

input ModelConversationMessagepirateChatConditionInput {
  conversationSessionId: ModelIDInput
  sender: ModelConversationMessageSenderInput
  content: ModelStringInput
  context: ModelStringInput
  uiComponents: ModelStringInput
  and: [ModelConversationMessagepirateChatConditionInput]
  or: [ModelConversationMessagepirateChatConditionInput]
  not: ModelConversationMessagepirateChatConditionInput
  createdAt: ModelStringInput
  updatedAt: ModelStringInput
  owner: ModelStringInput
}

input CreateConversationMessagepirateChatInput {
  id: ID
  conversationSessionId: ID!
  sender: ConversationMessageSender
  content: String
  context: AWSJSON
  uiComponents: [AWSJSON]
}

input DeleteConversationMessagepirateChatInput {
  id: ID!
}

input ModelSubscriptionConversationMessagepirateChatFilterInput {
  id: ModelSubscriptionIDInput
  conversationSessionId: ModelSubscriptionIDInput
  sender: ModelSubscriptionStringInput
  content: ModelSubscriptionStringInput
  context: ModelSubscriptionStringInput
  uiComponents: ModelSubscriptionStringInput
  createdAt: ModelSubscriptionStringInput
  updatedAt: ModelSubscriptionStringInput
  and: [ModelSubscriptionConversationMessagepirateChatFilterInput]
  or: [ModelSubscriptionConversationMessagepirateChatFilterInput]
  owner: ModelStringInput
}
"
`;
